### PR TITLE
Fixes #29082 -- Encodes JSON appropriately in the test Client.

### PR DESCRIPTION
**Pull request by NDevox**
Added option to switch the encoder if desired - although it has been pointed out that we could just subclass the Client instead.

Worth noting there are no additional tests for the PUT, PATCH and DELETE methods right now.

I am happy to add these I just don't have time _right now_, maybe tonight.

For now I'd like to get this reviewed to see if there is anything else we need to change.

### CODE REVIEW

**Comment by NDevox** [comment]
You can tell I'm developing in 3.6...

**Comment by NDevox** [comment]
@Cito I seemed to have swallowed your commit into my PR. Not sure how that happened?

**Comment by Cito** [comment]
Not sure how you did that. You probably need to rebase your PR branch onto the current master to fix it.

**Comment by NDevox** [comment]
Merged master in

**Comment by carltongibson** [comment]
@NDevox Any chance you could squash to a single commit with the canonical commit message?

**Comment by NDevox** [comment]
The tree on this one is a bit weird (still don't know why I've absorbed the commit that isn't mine). I also find squashing with a merge commit in the middle makes things a lot harder. Unless you know a good flow to get around it @carltongibson ?

**Comment by carltongibson** [comment]
OK. No problem. Let me see if I can clear it up for you.

**Comment by carltongibson** [comment]
OK. Just waiting for Jenkins (again) and I think we're good to go.

**Comment by NDevox** [comment]
can you share how you squashed it, if it was simple?

I always end up spending too much time squashing with merge commits.

**Comment by carltongibson** [comment]
I used `git rebase origin/master` — which gets your commits at the front — and then `git rebase -i` to squash the commits. 

I was just looking for a good link for you. [This StackOverflow post covers it nicely].(https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git) (I'll be using the highly upvoted `git reset --soft HEAD~3 && git commit` method — that's handy.)

**Comment by NDevox** [comment]
See I've normally used the first answer. I like soft reset too. Thanks!

File: django/test/client.py [hunk 1]
```diff
@@ -518,7 +538,10 @@ def get(self, path, data=None, follow=False, secure=False, **extra):
     def post(self, path, data=None, content_type=MULTIPART_CONTENT,
              follow=False, secure=False, **extra):
         """Request a response from the server using POST."""
-        response = super().post(path, data=data, content_type=content_type, secure=secure, **extra)
+        response = super().post(
+            path, data=data, content_type=content_type,
+            secure=secure, **extra
+        )
```
Comment (felixxm) [inline django/test/client.py]: I think you should revert this unrelated change, and also similar changes in `put()`, `patch()`, and `delete()`.

Comment (NDevox) [reply django/test/client.py]: This is an artifact from an earlier draft.

I am happy to revert it if desired. @carltongibson any opinion?

File: tests/test_client/tests.py [hunk 2]
```diff
@@ -85,6 +87,83 @@ def test_post(self):
         self.assertEqual(response.templates[0].name, 'POST Template')
         self.assertContains(response, 'Data received')
 
+    def test_json_post(self):
+        """ POST some JSON data to a view """
+        post_data = {
+            'value': 37
+        }
```
Comment (felixxm) [inline tests/test_client/tests.py]: This (and similar data dicts below) can be single lined.

File: tests/test_client/tests.py [hunk 3]
```diff
@@ -85,6 +87,83 @@ def test_post(self):
...
+        response = self.client.delete('/json_view/', delete_data, content_type=JSON_CONTENT)
+
+        # Check some response details
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.context['data'], 37)
+        self.assertEqual(response.templates[0].name, 'DELETE Template')
+        self.assertContains(response, 'Viewing DELETE page.')
+
+    def test_encoded_json_post(self):
```
Comment (felixxm) [inline tests/test_client/tests.py]: IMO you can combine this test with `test_json_post()` by using `subTest()`, e.g.:
```python
for post_data in ({'value': 37}, '{"value": 37}'):
    with self.subTest(post_data=post_data):
        response = self.client.post('/json_view/', post_data, content_type=JSON_CONTENT)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context['data'], 37)
        self.assertEqual(response.templates[0].name, 'POST Template')
        self.assertContains(response, 'Viewing POST page.')
```

File: tests/test_client/tests.py [hunk 4]
```diff
@@ -85,6 +87,83 @@ def test_post(self):
         self.assertEqual(response.templates[0].name, 'POST Template')
         self.assertContains(response, 'Data received')
 
+    def test_json_post(self):
+        """ POST some JSON data to a view """
```
Comment (felixxm) [inline tests/test_client/tests.py]: I think we prefer docstrings without leading and trailing spaces, i.e.
_"""POST JSON data to a view."""_.

File: tests/test_client/tests.py [hunk 5]
```diff
@@ -85,6 +87,83 @@ def test_post(self):
...
+        post_data = {
+            'value': 37
+        }
+        response = self.client.post('/json_view/', post_data, content_type=JSON_CONTENT)
+
+        # Check some response details
```
Comment (felixxm) [inline tests/test_client/tests.py]: I don't see much value in this comment. IMO it can be removed from all tests.

File: django/test/client.py (hunk 1 above)
Comment (carltongibson) [reply django/test/client.py]: Yes. Good spot @felixxm. 

I think we try and keep the change footprints as small as possible. It's unrelated so best to leave it out. 

(We could do it in a separate change, but the original is not so long as to offend — to my eye at least...)

File: tests/test_client/tests.py (hunk 5 above)
Comment (NDevox) [reply tests/test_client/tests.py]: Copy paste, will remove.

File: tests/test_client/tests.py (hunk 2 above)
Comment (NDevox) [reply tests/test_client/tests.py]: copy paste from similar test, will reformat.

File: tests/test_client/tests.py (hunk 3 above)
Comment (NDevox) [reply tests/test_client/tests.py]: They're testing slightly different things though. So I feel that warrants separate tests.

I also find subtest confuses things if there is a failure.

File: tests/test_client/tests.py (hunk 4 above)
Comment (NDevox) [reply tests/test_client/tests.py]: Is there a preference? Every doc string seems to have it's own formatting. I'm almost tempted to unify all of them. But that would be a very different ticket.

**Comment by NDevox** [comment]
@felixxm some of the changes, dict format, comment style, useless comment, come from other tests which I (lazily) didn't remove. I've edited them out of my tests. But I've also removed them from the pre-existing tests.

Want me to leave the old tests as they are or happy with that?

File: django/test/client.py (hunk 1 above)
Comment (NDevox) [reply django/test/client.py]: I wouldn't say it warrants a change. It stays under 120 chars so in my mind is fine.

**Comment by NDevox** [comment]
Right I did a --soft squash and have seemingly taken in more changes as a result to do with the cache. No idea why.

**Comment by felixxm** [comment]
@NDevox You should revert all unrelated changes, because we don't want to mix cleanups and refactoring with a real bug fix.

**Comment by NDevox** [comment]
Don't think I can cleanly revert at this point as the commits are pretty intertwined (code that isn't mine mixed with my own in a single commit). Unless anyone has a better idea I'm going to create a new branch off django/master and put up a new PR

**Comment by adamchainz** [comment]
You can always fix git branches locally then force push to overwrite :)

**Comment by NDevox** [comment]
Honestly I did try. I gave up after it said I had commited the last 15 commits.

Looks like an issue with force rebasing that I don't fully understand yet.
